### PR TITLE
System Tests: Module (FIM) - FIM Files Events: System test for FIM synchronization process

### PR DESCRIPTION
**Pull request by CamiRomero**
|Related issue|
|---|
[2434](https://github.com/wazuh/wazuh-qa/issues/2434)

## Description
While developing wazuh/wazuh#9103 we are planning to develop some system tests to check that the FIM database agent and the manager database are synchronized.

The changes made are as follows:

- Added new script to performe query to DB - path: tests/system/scripts/system_query_db.py
- Added  new function query_db in __init__.py to can executed scripts to DB inside the host 
- Added tests/system/test_fim/test_synchronization/data/delete_message.ymlto check expected outputs
- Added tests/system/test_fim/test_synchronization/data/messages.yml to check expected outputs
- Added tests/system/test_fim/test_synchronization/data/wait_fim_scan.yml to check expected outputs
- Added tests/system/test_fim/test_synchronization/data/agent_initializing_synchronization.yml to check expected outputs
- Added tests/system/test_fim/test_synchronization/data/manager_initializing_synchronization.yml to check expected outputs
- Added tests/system/test_fim/test_synchronization/test_synchronization.py to performance de test described in [2434](https://github.com/wazuh/wazuh-qa/issues/2434)

## Configuration options
In order to run the test, first the environment located in /test/system/provisioning/one_manager_agent must be enabled with:
sudo ansible-playbook -i inventory.yml playbook.yml --extra-vars='{"wazuh_branch": "4.2"}'

## Logs example

|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7981141/R1.zip)| 01/02/2022 | @CamiRomero  | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7981142/R2.zip) | 01/02/2022 | @CamiRomero   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7981143/R3.zip)| 01/02/2022 | @CamiRomero   | 
## Tests

- [x] Proven that tests **pass** when they have to pass.
- [x] Proven that tests **fail** when they have to fail.

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.

### CODE REVIEW

File: tests/system/test_fim/test_files/test_files_cud.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import pytest
+from wazuh_testing.tools.monitoring import HostMonitor
+from wazuh_testing.tools.system import HostManager
+from test_fim import create_folder_file, wait_for_fim_scan_end, clean_logs
+
+
+# Hosts
+testinfra_hosts = ["wazuh-manager", "wazuh-agent1"]
+
```
Comment (damarisg) [inline tests/system/test_fim/test_files/test_files_cud.py:53]: When do you use it?

File: tests/system/test_fim/test_synchronization/test_synchronization.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+from time import sleep
+
+
+from wazuh_testing.tools.monitoring import HostMonitor
+from wazuh_testing.tools.system import HostManager
+from test_fim import create_folder_file, clean_logs, query_db
+
+
+# Hosts
+testinfra_hosts = ["wazuh-manager", "wazuh-agent1"]
```
Comment (damarisg) [inline tests/system/test_fim/test_synchronization/test_synchronization.py:55]: When do you use it?

File: tests/system/test_fim/test_synchronization/test_synchronization.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    try:
+        HostMonitor(inventory_path=inventory_path,
+                    messages_path=messages_path[0],
+                    tmp_path=tmp_path).run()
+    except:
+        host_manager.run_command('wazuh-agent1', f'rm -rf {folder_path}')
+
+    clean_logs(host_manager)
+
+    # Stop agent
```
Comment (damarisg) [inline tests/system/test_fim/test_synchronization/test_synchronization.py:123]: There are 2 cases: the agent stops and the manager stops.

File: tests/system/test_fim/test_synchronization/test_synchronization.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+tmp_path = os.path.join(local_path, 'tmp')
+scheduled_mode = 'testdir1'
+db_path = '/var/ossec/queue/db/001.db'
+db_script = '/var/system_query_db.py'
+
+
+@pytest.mark.parametrize('host', ['wazuh-agent1', 'wazuh-manager'])
+@pytest.mark.parametrize('case', ['add', 'modify', 'delete'])
+@pytest.mark.parametrize('folder_path', ['testdir1'])
+def test_Synchronization_create_file_agent_stopped(folder_path, case, host):
```
Comment (damarisg) [inline tests/system/test_fim/test_synchronization/test_synchronization.py:76]: The name should be representative. It doesn't just create a file (it also modifies and deletes it) and, the description says that it is when the manager or agent stops (2 cases). However, the test name only talks about agent stopped.

**Review by damarisg — CHANGES_REQUESTED** [review]

**Comment by Deblintrake09** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7989761/R1.zip)| 02/02/2022 | @Deblintrake09   | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7989762/R2.zip) | 02/02/2022 | @Deblintrake09   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7989797/R3.zip)| 02/02/2022 | @Deblintrake09    |

File: tests/system/test_fim/test_synchronization/test_synchronization.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
+"""
+copyright: Copyright (C) 2015-2021, Wazuh Inc.
+           Created by Wazuh, Inc. <info@wazuh.com>.
+           This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+type: system
+brief: Check that when FIM is activated, and the agent is running, the agent and manager are synchronization when
+a change is performed in a monitored folder.
+tier: 1
```
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [83482](https://github.com/wazuh/wazuh-qa/pull/2536/commits/83482f6bae85d292711b62dccf44af32ba322218)

File: tests/system/test_fim/test_synchronization/test_synchronization.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+
+@pytest.mark.parametrize('host', ['wazuh-agent1', 'wazuh-manager'])
+@pytest.mark.parametrize('case', ['add', 'modify', 'delete'])
+@pytest.mark.parametrize('folder_path', ['testdir1'])
+def test_Synchronization_create_file_agent_stopped(folder_path, case, host):
+    '''
+    Description: The test will monitor a directory and apply changes when agent is stopped.
+    Finally, it will verify that the FIM 'Synchronization' event is generated
+    in agent and manager side.
```
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [71e04](https://github.com/wazuh/wazuh-qa/pull/2536/commits/71e04506cce02c49fe7931bfa6c8973ac6a9a474l)

File: tests/system/test_fim/test_synchronization/test_synchronization.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [7ae07](https://github.com/wazuh/wazuh-qa/pull/2536/commits/7ae0720f1da0c910bc5ba364481db18d525d6294)

Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [7ae07](https://github.com/wazuh/wazuh-qa/pull/2536/commits/7ae0720f1da0c910bc5ba364481db18d525d6294)

File: tests/system/test_fim/test_synchronization/test_synchronization.py (hunk 3 above)
Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [1321a](https://github.com/wazuh/wazuh-qa/pull/2536/commits/1321a0318f984b93edcdab0ec1d3ec9854c420f4)

File: tests/system/test_fim/test_synchronization/test_synchronization.py (hunk 2 above)
Comment (CamiRomero) [reply tests/system/test_fim/test_synchronization/test_synchronization.py]: Done in commit [080c1](https://github.com/wazuh/wazuh-qa/pull/2536/commits/080c12ab527c54c7dbed0a7b22af7c2cf4c668e0)

File: tests/system/test_fim/test_files/test_files_cud.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
+"""
+copyright: Copyright (C) 2015-2021, Wazuh Inc.
+           Created by Wazuh, Inc. <info@wazuh.com>.
+           This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+type: system
+brief: Check that when FIM is activated, and the agent is running, the agent and manager are synchronization when
+a change is performed in a monitored folder.
+tier: 1
```
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_files/test_files_cud.py:8]: This is not part of this PR. I applied the change in [2443](https://github.com/wazuh/wazuh-qa/pull/2443)

File: tests/system/test_fim/test_files/test_files_cud.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    - Red Hat 6
+references:
+    - https://github.com/wazuh/wazuh-qa/issues/2389
+tags:
+    - fim_basic_usage
+"""
+
+import os
+
+import pytest
```
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_files/test_files_cud.py:45]: This is not part of this PR. I applied the change in https://github.com/wazuh/wazuh-qa/pull/2443

File: tests/system/test_fim/test_files/test_files_cud.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+
+@pytest.mark.parametrize('case', ['add', 'modify', 'delete'])
+@pytest.mark.parametrize('folder_path', ['testdir1', 'testdir2', 'testdir3'])
+def test_file_cud(folder_path, case):
+    '''
+    The test will monitor a directory.
+    Finally, it will verify that the FIM event is generated
+    in agent and manager side.
+    '''
```
[the remark this answers is not in the document]
Comment (CamiRomero) [reply tests/system/test_fim/test_files/test_files_cud.py:72]: This was worked in PR [2443](This is not part of this PR. [2443](https://github.com/wazuh/wazuh-qa/pull/2443)

File: tests/system/test_fim/__init__.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+                tmp_path=tmp_path).run()
+
+
+def clean_logs(host_manager):
+    host_manager.clear_file(host='wazuh-manager', file_path=os.path.join(WAZUH_LOGS_PATH, 'ossec.log'))
+    host_manager.clear_file(host='wazuh-agent1', file_path=os.path.join(WAZUH_LOGS_PATH, 'ossec.log'))
+
+
+def query_db(host_manager, script, db_path, query):
+    return host_manager.run_command('wazuh-manager', "python {} --db_path {} --query {}".format(script, db_path, query))
```
Comment (damarisg) [inline tests/system/test_fim/__init__.py:28]: Comments are necessary for each method. Why don't you have them here?

**Review by damarisg — CHANGES_REQUESTED** [review]

**Comment by CamiRomero** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7996596/R1.zip)| 03/02/2022 | @CamiRomero  | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7996599/R2.zip) | 03/02/2022 | @CamiRomero   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7996600/R3.zip)| 03/02/2022 | @CamiRomero   |

**Comment by Deblintrake09** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/7997528/R1.zip)| 03/02/2022 | @Deblintrake09   | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7997526/R2.zip) | 03/02/2022 | @Deblintrake09   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7997527/R3.zip)| 03/02/2022 | @Deblintrake09    |

**Review by Deblintrake09 — APPROVED** [review]

**Comment by CamiRomero** [comment]
|Run| Results | Date | By | Notes |
|--|--|--|--|--|
| R1 | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8050793/R1.zip)| 03/02/2022 | @CamiRomero  | 
| R2 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8050804/R2.zip)| 03/02/2022 | @CamiRomero   | 
| R3 | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8050805/R3.zip)| 03/02/2022 | @CamiRomero   |

**Pull request merged**